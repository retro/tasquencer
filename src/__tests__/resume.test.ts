import { Effect } from 'effect';
import { expect, it } from 'vitest';

import { Builder, IdGenerator, Service } from '../index.js';
import { getEnabledTaskNames, makeIdGenerator } from './shared.js';

const subWorkflowDefinition = Builder.workflow()
  .withName('sub')
  .withParentContext()
  .startCondition('start')
  .task('subT1')
  .endCondition('end')
  .connectCondition('start', (to) => to.task('subT1'))
  .connectTask('subT1', (to) => to.condition('end'));

const workflowDefinition = Builder.workflow()
  .withName('activities')
  .startCondition('start')
  .task('t1', Builder.emptyTask().withSplitType('and'))
  .task('t2')
  .compositeTask('t3', (ct) => ct().withSubWorkflow(subWorkflowDefinition))
  .task('t4', Builder.emptyTask().withJoinType('and'))
  .endCondition('end')
  .connectCondition('start', (to) => to.task('t1'))
  .connectTask('t1', (to) => to.task('t2').task('t3'))
  .connectTask('t2', (to) => to.task('t4'))
  .connectTask('t3', (to) => to.task('t4'))
  .connectTask('t4', (to) => to.condition('end'));

it('resumes from persistable state', () => {
  const program = Effect.gen(function* ($) {
    const idGenerator = makeIdGenerator();

    const service1 = yield* $(
      workflowDefinition.build(),
      Effect.flatMap((workflow) => Service.initialize(workflow)),
      Effect.provideService(IdGenerator, idGenerator)
    );

    yield* $(service1.start());
    const state1_1 = yield* $(service1.getState());
    expect(state1_1).toMatchSnapshot();
    expect(getEnabledTaskNames(state1_1)).toEqual(new Set(['t1']));

    yield* $(service1.startTask('t1'));
    const state1_2 = yield* $(service1.getState());
    expect(state1_2).toMatchSnapshot();
    expect(getEnabledTaskNames(state1_2)).toEqual(new Set(['t2', 't3']));

    yield* $(service1.startTask('t2'));
    const state1_3 = yield* $(service1.getState());
    expect(state1_3).toMatchSnapshot();
    expect(getEnabledTaskNames(state1_3)).toEqual(new Set(['t3']));

    const service2 = yield* $(
      workflowDefinition.build(),
      Effect.flatMap((workflow) => Service.resume(workflow, state1_3)),
      Effect.provideService(IdGenerator, idGenerator)
    );

    const state2_1 = yield* $(service2.getState());
    expect(state2_1).toEqual(state1_3);

    yield* $(service2.startTask('t3'));
    const state2_2 = yield* $(service2.getState());
    expect(state2_2).toMatchSnapshot();
    expect(getEnabledTaskNames(state2_2)).toEqual(new Set());

    const { id: parentWorkItemId } = yield* $(
      service2.initializeWorkItem('t2')
    );
    const { id: subWorkflowId } = yield* $(service2.initializeWorkflow('t3'));
    yield* $(service2.startWorkflow(`t3.${subWorkflowId}`));
    yield* $(service2.startTask(`t3.${subWorkflowId}.subT1`));
    const { id: workItemId } = yield* $(
      service2.initializeWorkItem(`t3.${subWorkflowId}.subT1`)
    );
    yield* $(service2.startWorkItem(`t2.${parentWorkItemId}`));
    yield* $(service2.startWorkItem(`t3.${subWorkflowId}.subT1.${workItemId}`));
    yield* $(
      service2.completeWorkItem(`t3.${subWorkflowId}.subT1.${workItemId}`)
    );
    yield* $(service2.completeWorkItem(`t2.${parentWorkItemId}`));
    const state2_3 = yield* $(service2.getState());
    expect(state2_3).toMatchSnapshot();
    expect(getEnabledTaskNames(state2_3)).toEqual(new Set(['t4']));

    yield* $(service2.startTask('t4'));
    const state2_4 = yield* $(service2.getState());
    expect(state2_4).toMatchSnapshot();
    expect(getEnabledTaskNames(state2_4)).toEqual(new Set());
    expect(state2_4.workflows[0]?.state).toEqual('completed');
  });

  Effect.runSync(program);
});
