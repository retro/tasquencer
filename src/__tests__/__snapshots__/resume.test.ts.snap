// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`resumes from persistable state 1`] = `
{
  "conditions": [
    {
      "marking": 1,
      "name": "start",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "end",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "implicit:t1->t2",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "implicit:t1->t3",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "implicit:t2->t4",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "implicit:t3->t4",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
  ],
  "tasks": [
    {
      "generation": 0,
      "name": "t1",
      "state": "enabled",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "generation": 0,
      "name": "t2",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "generation": 0,
      "name": "t3",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "generation": 0,
      "name": "t4",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
  ],
  "workItems": [],
  "workflows": [
    {
      "context": undefined,
      "id": "workflow-1",
      "name": "activities",
      "parent": null,
      "state": "started",
    },
  ],
}
`;

exports[`resumes from persistable state 2`] = `
{
  "conditions": [
    {
      "marking": 0,
      "name": "start",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "end",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 1,
      "name": "implicit:t1->t2",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 1,
      "name": "implicit:t1->t3",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "implicit:t2->t4",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "implicit:t3->t4",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
  ],
  "tasks": [
    {
      "generation": 1,
      "name": "t1",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "generation": 0,
      "name": "t2",
      "state": "enabled",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "generation": 0,
      "name": "t3",
      "state": "enabled",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "generation": 0,
      "name": "t4",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
  ],
  "workItems": [],
  "workflows": [
    {
      "context": undefined,
      "id": "workflow-1",
      "name": "activities",
      "parent": null,
      "state": "started",
    },
  ],
}
`;

exports[`resumes from persistable state 3`] = `
{
  "conditions": [
    {
      "marking": 0,
      "name": "start",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "end",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "implicit:t1->t2",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 1,
      "name": "implicit:t1->t3",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "implicit:t2->t4",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "implicit:t3->t4",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
  ],
  "tasks": [
    {
      "generation": 1,
      "name": "t1",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "generation": 1,
      "name": "t2",
      "state": "started",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "generation": 0,
      "name": "t3",
      "state": "enabled",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "generation": 0,
      "name": "t4",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
  ],
  "workItems": [],
  "workflows": [
    {
      "context": undefined,
      "id": "workflow-1",
      "name": "activities",
      "parent": null,
      "state": "started",
    },
  ],
}
`;

exports[`resumes from persistable state 4`] = `
{
  "conditions": [
    {
      "marking": 0,
      "name": "start",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "end",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "implicit:t1->t2",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "implicit:t1->t3",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "implicit:t2->t4",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "implicit:t3->t4",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
  ],
  "tasks": [
    {
      "generation": 1,
      "name": "t1",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "generation": 1,
      "name": "t2",
      "state": "started",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "generation": 1,
      "name": "t3",
      "state": "started",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "generation": 0,
      "name": "t4",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
  ],
  "workItems": [],
  "workflows": [
    {
      "context": undefined,
      "id": "workflow-1",
      "name": "activities",
      "parent": null,
      "state": "started",
    },
  ],
}
`;

exports[`resumes from persistable state 5`] = `
{
  "conditions": [
    {
      "marking": 0,
      "name": "start",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "end",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "implicit:t1->t2",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "implicit:t1->t3",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 1,
      "name": "implicit:t2->t4",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 1,
      "name": "implicit:t3->t4",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "start",
      "workflowId": "workflow-2",
      "workflowName": "sub",
    },
    {
      "marking": 1,
      "name": "end",
      "workflowId": "workflow-2",
      "workflowName": "sub",
    },
  ],
  "tasks": [
    {
      "generation": 1,
      "name": "t1",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "generation": 1,
      "name": "t2",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "generation": 1,
      "name": "t3",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "generation": 0,
      "name": "t4",
      "state": "enabled",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "generation": 1,
      "name": "subT1",
      "state": "completed",
      "workflowId": "workflow-2",
      "workflowName": "sub",
    },
  ],
  "workItems": [
    {
      "id": "workItem-1",
      "payload": null,
      "state": "completed",
      "taskGeneration": 1,
      "taskName": "t2",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "id": "workItem-2",
      "payload": null,
      "state": "completed",
      "taskGeneration": 1,
      "taskName": "subT1",
      "workflowId": "workflow-2",
      "workflowName": "sub",
    },
  ],
  "workflows": [
    {
      "context": undefined,
      "id": "workflow-1",
      "name": "activities",
      "parent": null,
      "state": "started",
    },
    {
      "context": undefined,
      "id": "workflow-2",
      "name": "sub",
      "parent": {
        "taskGeneration": 1,
        "taskName": "t3",
        "workflowId": "workflow-1",
        "workflowName": "activities",
      },
      "state": "completed",
    },
  ],
}
`;

exports[`resumes from persistable state 6`] = `
{
  "conditions": [
    {
      "marking": 0,
      "name": "start",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 1,
      "name": "end",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "implicit:t1->t2",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "implicit:t1->t3",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "implicit:t2->t4",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "implicit:t3->t4",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "start",
      "workflowId": "workflow-2",
      "workflowName": "sub",
    },
    {
      "marking": 1,
      "name": "end",
      "workflowId": "workflow-2",
      "workflowName": "sub",
    },
  ],
  "tasks": [
    {
      "generation": 1,
      "name": "t1",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "generation": 1,
      "name": "t2",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "generation": 1,
      "name": "t3",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "generation": 1,
      "name": "t4",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "generation": 1,
      "name": "subT1",
      "state": "completed",
      "workflowId": "workflow-2",
      "workflowName": "sub",
    },
  ],
  "workItems": [
    {
      "id": "workItem-1",
      "payload": null,
      "state": "completed",
      "taskGeneration": 1,
      "taskName": "t2",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "id": "workItem-2",
      "payload": null,
      "state": "completed",
      "taskGeneration": 1,
      "taskName": "subT1",
      "workflowId": "workflow-2",
      "workflowName": "sub",
    },
  ],
  "workflows": [
    {
      "context": undefined,
      "id": "workflow-1",
      "name": "activities",
      "parent": null,
      "state": "completed",
    },
    {
      "context": undefined,
      "id": "workflow-2",
      "name": "sub",
      "parent": {
        "taskGeneration": 1,
        "taskName": "t3",
        "workflowId": "workflow-1",
        "workflowName": "activities",
      },
      "state": "completed",
    },
  ],
}
`;
